### PR TITLE
Enhancement: Use compute to offload heavy work from the main thread

### LINKED ISSUE #536: Improve `FileSystemTransport` by using `compute()` or background isolates
We can try to use [`compute()`](https://api.flutter.dev/flutter/foundation/compute-constant.html) here

https://github.com/getsentry/sentry-dart/blob/bca59abc7f5f257b4fcbb973090d9df719509ebc/flutter/lib/src/file_system_transport.dart#L14-L17

in order to offload the conversion to another isolate.
This could improve the performance because we're not blocking the UI Isolate, especially if there are any attachments or big events.

There are [some things](https://api.flutter.dev/flutter/dart-isolate/SendPort/send.html) which need to be considered.

```dart
class NewFileSystemTransport implements Transport {
  NewFileSystemTransport(this._channel, this._options);

  final MethodChannel _channel;
  final SentryOptions _options;

  @override
  Future<SentryId?> send(SentryEnvelope envelope) async {
    final args = await compute(_convert, envelope);
    try {
      await _channel.invokeMethod<void>('captureEnvelope', [args]);
    } catch (exception, stackTrace) {
      _options.logger(
        SentryLevel.error,
        'Failed to save envelope',
        exception: exception,
        stackTrace: stackTrace,
      );
      return SentryId.empty();
    }

    return envelope.header.eventId;
  }

  static Future<Uint8List> _convert(SentryEnvelope envelope) async {
    final envelopeData = <int>[];
    await envelope.envelopeStream().forEach(envelopeData.addAll);
    // https://flutter.dev/docs/development/platform-integration/platform-channels#codec
    return Uint8List.fromList(envelopeData);
  }
}
```

### PR DESCRIPTION
**Pull request by marandaneto**
## :scroll: Description
Enhancement: Use compute to offload heavy work from the main thread

## :bulb: Motivation and Context
Closes https://github.com/getsentry/sentry-dart/issues/536

## :pencil: Checklist

- [X] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] I updated the docs if needed
- [X] All tests passing
- [X] No breaking changes

### CODE REVIEW

**Comment by ueman** [comment]
Does it work for file attachments?

**Comment by marandaneto** [comment]
getting the error:

> [sentry] [error] Error while capturing exception
> [sentry] ArgumentError (Invalid argument(s): Illegal argument in isolate message: (object is aReceivePort))
> [sentry] #0      Isolate._spawnFunction (dart:isolate-patch/isolate_patch.dart:395:25)
> #1      Isolate.spawn (dart:isolate-patch/isolate_patch.dart:375:7)
> #2      compute
> package:flutter/…/foundation/_isolates_io.dart:22
> #3      FileSystemTransport.send
> package:sentry_flutter/src/file_system_transport.dart:16

**Comment by marandaneto** [comment]
> Does it work for file attachments?

Didn't test it yet, not working even for Events right now.
I dare to say its related to the limited data types accepted by https://api.dart.dev/stable/2.14.4/dart-isolate/SendPort-class.html

File: flutter/lib/src/file_system_transport.dart [hunk 1]
```diff
@@ -11,10 +12,12 @@ class FileSystemTransport implements Transport {
...
-    // https://flutter.dev/docs/development/platform-integration/platform-channels#codec
-    final args = [Uint8List.fromList(envelopeData)];
+    final eventIdLabel = envelope.header.eventId?.toString() ?? '';
+    final args = await compute(
+      _convert,
+      _CaptureEnvelopeData(envelope, _options),
+      debugLabel: 'captureEnvelope $eventIdLabel',
+    );
     try {
       await _channel.invokeMethod<void>('captureEnvelope', args);
```
Comment (ueman) [inline flutter/lib/src/file_system_transport.dart]: `args` previously was an array 
```suggestion
      await _channel.invokeMethod<void>('captureEnvelope', [args]);
```

Comment (ueman) [reply flutter/lib/src/file_system_transport.dart]: There should be a comment which tells you which type should get passed. Probably also a test which asserts that 😅

Comment (marandaneto) [reply flutter/lib/src/file_system_transport.dart]: that's indeed another problem, but the error is actually with the `compute` flag, See https://github.com/getsentry/sentry-dart/pull/707#issuecomment-1015772386

Comment (ueman) [reply flutter/lib/src/file_system_transport.dart]: Well, at least the tests are now passing.

**Pull request closed without merging**